src/Mvc/Mvc.Analyzers/src/TagHelpersInCodeBlocksAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class TagHelpersInCodeBlocksAnalyzer : DiagnosticAnalyzer
+    {
+        public TagHelpersInCodeBlocksAnalyzer()
+        {
+            TagHelperInCodeBlockDiagnostic = DiagnosticDescriptors.MVC1006_FunctionsContainingTagHelpersMustBeAsyncAndReturnTask;
+            SupportedDiagnostics = ImmutableArray.Create(new[] { TagHelperInCodeBlockDiagnostic });
+        }
+
+        private DiagnosticDescriptor TagHelperInCodeBlockDiagnostic { get; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                var symbolCache = new SymbolCache(context.Compilation);
```

Comment:
I am trying out a `TryCreate` pattern inspired by a different PR that @rynowak sent out: (https://github.com/aspnet/AspNetCore/pull/10375/files#diff-873a6b7499a94ad1cace9d0ecb095fa2R26). Might be useful to consider switching this here.

Reply:
```C#
if (!SymbolCache.TryCreate(context.Compilation, out var symbolCache))
{
                    // No-op if we can't find types we care about.
                    return;
```